src/thor/trace_route_action.cc

```diff
@@ -419,34 +423,61 @@ thor_worker_t::map_match(Api& request) {
             discontinuity_edges_iter->first == prev_path_edge_itr->edgeid &&
             discontinuity_edges_iter->second == path_edge_itr->edgeid) {
           // start a new one
-          edge_groups.emplace_back(edge_group_t{path_edge_itr, path_edge_itr, {}, {}});
+          edge_groups.emplace_back(edge_group_t{path_edge_itr, path_edge_itr, {}, {}, false});
+          idx = 1;
+          edge_table.clear();
           ++discontinuity_edges_iter;
         }
         // remember where the last one ended
         std::get<1>(edge_groups.back()) = path_edge_itr;
+        // also remember whether there is a loop on the last one (gap in loop idx)
+        int loop_idx = edge_table[path_edge_itr->edgeid];
+        if (loop_idx > 0 && loop_idx != idx - 1) {
+          std::get<4>(edge_groups.back()) = true;
+        } else {
+          std::get<4>(edge_groups.back()) = false;
+          edge_table[path_edge_itr->edgeid] = ++idx;
+        }
       }
 
       // here we mark the original shape points that just after and just before the
       // discontinuities so that in the path they get "upgraded" to break points
       auto match_result_itr = match_results.cbegin();
       for (auto& edge_group : edge_groups) {
-        // we know the discontinuity happens on this edge so we just need to
-        // find the first match result on this edge
-        auto first_edge_id = std::get<0>(edge_group)->edgeid;
-        while (match_result_itr != match_results.cend() &&
-               match_result_itr->edgeid != first_edge_id) {
-          ++match_result_itr;
-        }
+        auto first_edge = std::get<0>(edge_group);
+        auto last_edge = std::get<1>(edge_group);
+
+        match_result_itr = std::find_if(match_result_itr, match_results.cend(),
+                                        [first_edge](const MatchResult& result) {
+                                          return result.edgeid == first_edge->edgeid;
+                                        });
```

Comment:
if this code is equivalent to the previous code, A why change it, and B why remove the comments? can you give some explanation of the changes here?

Reply:
To make consistency with the later code edited using std::find to locate where the discontinuity begins. And, I shouldn't delete the comments. Added it back.